From: Chris Wright <chrisw@sous-sol.org>
Subject: [PATCH] pci: check caps from sysfs file open to read device dependent config space
Git-commit: de139a3393958c5adc75b7df7619d7e48d9ea559
Patch-mainline: 2.6.35
Reference: bnc#706131

The PCI config space bin_attr read handler has a hardcoded CAP_SYS_ADMIN
check to verify privileges before allowing a user to read device
dependent config space.  This is meant to protect from an unprivileged
user potentially locking up the box.

When assigning a PCI device directly to a guest with libvirt and KVM,
the sysfs config space file is chown'd to the unprivileged user that
the KVM guest will run as.  The guest needs to have full access to the
device's config space since it's responsible for driving the device.
However, despite being the owner of the sysfs file, the CAP_SYS_ADMIN
check will not allow read access beyond the config header.

With this patch we check privileges against the capabilities used when
openining the sysfs file.  The allows a privileged process to open the
file and hand it to an unprivileged process, and the unprivileged process
can still read all of the config space.

Signed-off-by: Chris Wright <chrisw@sous-sol.org>
Acked-by: Jesse Barnes <jbarnes@virtuousgeek.org>
Cc: Alan Cox <alan@lxorguk.ukuu.org.uk>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>
Acked-by: Bruce Rogers <brogers@suse.com>
---
 drivers/pci/pci-sysfs.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

Index: kernel/drivers/pci/pci-sysfs.c
===================================================================
--- kernel.orig/drivers/pci/pci-sysfs.c
+++ kernel/drivers/pci/pci-sysfs.c
@@ -21,6 +21,7 @@
 #include <linux/stat.h>
 #include <linux/topology.h>
 #include <linux/mm.h>
+#include <linux/fs.h>
 #include <linux/capability.h>
 #include <linux/pci-aspm.h>
 #include "pci.h"
@@ -339,7 +340,7 @@ pci_read_config(struct file *filp, struc
 	u8 *data = (u8*) buf;
 
 	/* Several chips lock up trying to read undefined config space */
-	if (capable(CAP_SYS_ADMIN)) {
+	if (cap_raised(filp->f_cred->cap_effective, CAP_SYS_ADMIN)) {
 		size = dev->cfg_size;
 	} else if (dev->hdr_type == PCI_HEADER_TYPE_CARDBUS) {
 		size = 128;
Index: kernel/drivers/net/qlcnic/qlcnic_main.c
===================================================================
--- kernel.orig/drivers/net/qlcnic/qlcnic_main.c
+++ kernel/drivers/net/qlcnic/qlcnic_main.c
@@ -3631,7 +3631,8 @@ validate_pm_config(struct qlcnic_adapter
 }
 
 static ssize_t
-qlcnic_sysfs_write_pm_config(struct kobject *kobj, struct bin_attribute *attr,
+qlcnic_sysfs_write_pm_config(struct file *filp, struct kobject *kobj,
+			struct bin_attribute *attr,
 			char *buf, loff_t offset, size_t size)
 {
 	struct device *dev = container_of(kobj, struct device, kobj);
@@ -3670,7 +3671,8 @@ qlcnic_sysfs_write_pm_config(struct kobj
 }
 
 static ssize_t
-qlcnic_sysfs_read_pm_config(struct kobject *kobj, struct bin_attribute *attr,
+qlcnic_sysfs_read_pm_config(struct file *filp, struct kobject *kobj,
+			struct bin_attribute *attr,
 			char *buf, loff_t offset, size_t size)
 {
 	struct device *dev = container_of(kobj, struct device, kobj);
@@ -3861,7 +3863,8 @@ validate_npar_config(struct qlcnic_adapt
 }
 
 static ssize_t
-qlcnic_sysfs_write_npar_config(struct kobject *kobj, struct bin_attribute *attr,
+qlcnic_sysfs_write_npar_config(struct file *filp, struct kobject *kobj,
+		struct bin_attribute *attr,
 		char *buf, loff_t offset, size_t size)
 {
 	struct device *dev = container_of(kobj, struct device, kobj);
@@ -3900,7 +3903,8 @@ qlcnic_sysfs_write_npar_config(struct ko
 
 }
 static ssize_t
-qlcnic_sysfs_read_npar_config(struct kobject *kobj, struct bin_attribute *attr,
+qlcnic_sysfs_read_npar_config(struct file *filp, struct kobject *kobj,
+		struct bin_attribute *attr,
 		char *buf, loff_t offset, size_t size)
 {
 	struct device *dev = container_of(kobj, struct device, kobj);
@@ -3933,7 +3937,8 @@ qlcnic_sysfs_read_npar_config(struct kob
 }
 
 static ssize_t
-qlcnic_sysfs_get_port_stats(struct kobject *kobj, struct bin_attribute *attr,
+qlcnic_sysfs_get_port_stats(struct file *filp, struct kobject *kobj,
+		struct bin_attribute *attr,
 		char *buf, loff_t offset, size_t size)
 {
 	struct device *dev = container_of(kobj, struct device, kobj);
@@ -3963,7 +3968,8 @@ qlcnic_sysfs_get_port_stats(struct kobje
 }
 
 static ssize_t
-qlcnic_sysfs_get_esw_stats(struct kobject *kobj, struct bin_attribute *attr,
+qlcnic_sysfs_get_esw_stats(struct file *filp, struct kobject *kobj,
+		struct bin_attribute *attr,
 		char *buf, loff_t offset, size_t size)
 {
 	struct device *dev = container_of(kobj, struct device, kobj);
@@ -3993,7 +3999,8 @@ qlcnic_sysfs_get_esw_stats(struct kobjec
 }
 
 static ssize_t
-qlcnic_sysfs_clear_esw_stats(struct kobject *kobj, struct bin_attribute *attr,
+qlcnic_sysfs_clear_esw_stats(struct file *filp, struct kobject *kobj,
+		struct bin_attribute *attr,
 		char *buf, loff_t offset, size_t size)
 {
 	struct device *dev = container_of(kobj, struct device, kobj);
@@ -4017,7 +4024,8 @@ qlcnic_sysfs_clear_esw_stats(struct kobj
 }
 
 static ssize_t
-qlcnic_sysfs_clear_port_stats(struct kobject *kobj, struct bin_attribute *attr,
+qlcnic_sysfs_clear_port_stats(struct file *filp, struct kobject *kobj,
+		struct bin_attribute *attr,
 		char *buf, loff_t offset, size_t size)
 {
 
@@ -4042,7 +4050,8 @@ qlcnic_sysfs_clear_port_stats(struct kob
 }
 
 static ssize_t
-qlcnic_sysfs_read_pci_config(struct kobject *kobj, struct bin_attribute *attr,
+qlcnic_sysfs_read_pci_config(struct file *filp, struct kobject *kobj,
+		struct bin_attribute *attr,
 		char *buf, loff_t offset, size_t size)
 {
 	struct device *dev = container_of(kobj, struct device, kobj);
